### PR TITLE
Split loading and managing segments out of SegmentedJournal

### LINKED ISSUE #7604: Split the loading and creating of segments from SegmentedJournal
SegmentedJournal is responsible for too many things. We can split the loading and creating of the segments to another class. SegmentedJournal will be only an entry point to the journal.

The new class would know how to load a segment file and how to create a segment file.

Follow up #6221

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

Split loading and managing segments out of `SegmentedJournal`. The goal is not reduce the responsbility of `SegmentedJournal` class.

This PR only moved some code around. In addition, it also removed `synchronized` from a lot of methods. It seems unnecessary, as we already assume a single-writer multiple-readers semantics. Having `synchronized` is not helping us to prevent any race conditions. 

## Related issues

closes #7604 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: journal/src/main/java/io/camunda/zeebe/journal/file/SegmentFileManager.java
Comment (npepinpe) [inline journal/src/main/java/io/camunda/zeebe/journal/file/SegmentFileManager.java:27]: :wrench: From the functionality, it can: `loadExistingSegment` and `createSegment`. I think `SegmentLoader` is fine as a name, since tbh the `JournalSegments` is the one which "manages" segments (e.g. keeps track of them, handles corruption, deletion, etc.).

```suggestion
final class SegmentLoader {
```

File: journal/src/main/java/io/camunda/zeebe/journal/file/SegmentFileManager.java
Comment (npepinpe) [inline journal/src/main/java/io/camunda/zeebe/journal/file/SegmentFileManager.java:33]:
```suggestion
  SegmentFileManager(final long lastWrittenIndex, final JournalIndex journalIndex) {
```

File: journal/src/main/java/io/camunda/zeebe/journal/file/SegmentFileManager.java
Comment (npepinpe) [inline journal/src/main/java/io/camunda/zeebe/journal/file/SegmentFileManager.java:130]: :upside_down_face: I would group the "API" of this class together (i.e. non private functions), since private things are implementation details, so someone looking into this class only has to check the top of the file.

File: journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegments.java
Comment (npepinpe) [inline journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegments.java:64]: :wrench: I would rather inject the `SegmentFileManager` instead of constructing it. It might also make it easier to test users of this class (e.g. `SegmentedJournal`) if we can provide a different implementation (though I'd have to look if that's currently the case, I don't have all the tests OTOH now).

**Review by npepinpe — APPROVED** [review]
:rocket: 

I like it a lot! Generally most of these comments/changes don't need a second review, but feel free to request one if you'd like. Please address them before merging though (at least the suggestions/requested changes), even if it's to tell me how wrong I am :smile: 

- :thought_balloon: I wonder why we sometimes use `Segment` (or `Segmented`) as prefix, and sometimes not :smile: It's a total nitpick, not important in the end, but for some reason it sticks out to me.
- :x: Some of the tests in `SegmentedJournalTest` are now clearly for the `JournalSegments` class. Can we split the test class as well? This will reduce the "blast radius" of the tests, since the `JournalSegmentsTest` will be much more focused. For example, tests like `shouldHandlePartiallyWrittenDescriptor`, `shouldHandleCorruptionAtDescriptorWithoutAckedEntries`, and possibly other related ones.
- :x: If `JournalSegments` is meant to be used only within its own package, then we should make its usable API and itself package private, to avoid anybody ever using it outside. Same with other classes in general, e.g. `SegmentFileManager` (though I think that's already the case there).

File: journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegments.java
Comment (deepthidevaki) [reply journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegments.java:29]: It is not actually thread-safe. We expect that only there is only writer. But we don't enforce it. We have rwlocks only to synchronize between readers and the writer in specific operations. `append` for example doesn't use any synchronization because we assume that there is only one writer. We should either document this or enforce it as part of other issue. For now, it doesn't make sense to me that some methods are "synchronized" while some are not. That's why I removed it.

**Comment by deepthidevaki** [comment]
> * I wonder why we sometimes use `Segment` (or `Segmented`) as prefix, and sometimes not smile It's a total nitpick, not important in the end, but for some reason it sticks out to me.

The meaning is different `SegmentedJournal` - Segmented is adjective. `SegmentManager` - somthing that manages a segment.

> 
> * x Some of the tests in `SegmentedJournalTest` are now clearly for the `JournalSegments` class. Can we split the test class as well? This will reduce the "blast radius" of the tests, since the `JournalSegmentsTest` will be much more focused. For example, tests like `shouldHandlePartiallyWrittenDescriptor`, `shouldHandleCorruptionAtDescriptorWithoutAckedEntries`, and possibly other related ones.

I split some of the tests into a new class. Some tests remain in the `SegmentedJournalTest` because they are testing "integration", but added a new test in SegmentManagerTest that verifies a lower scope. 

I have addressed other review comments.

File: journal/src/test/java/io/camunda/zeebe/journal/file/SegmentsManagerTest.java
Comment (npepinpe) [inline journal/src/test/java/io/camunda/zeebe/journal/file/SegmentsManagerTest.java:124]: :+1: Uh, I didn't know this one! I only knew the `assertThatCode(() -> {}).doesNotThrowException()`. TIL

**Review by npepinpe — APPROVED** [review]

**Comment by deepthidevaki** [comment]
bors merge

**Pull request merged**
